reports/2017-05-26.md

```diff
@@ -0,0 +1,131 @@
+
+
+### Creating a client for Docker
+
+```go
+client, err := containerd.New(address, containerd.WithNamespace("docker"))
+if err != nil {
+	return err
+}
+defer client.Close()
+```
+
+### Pulling an Image from DockerHub
+
+```go
+// pull && unpack the image to your snapshot ( overlayfs default ) of choice
+image, err := client.Pull(ctx, "docker.io/library/redis:alpine", containerd.WithPullUnpack)
+if err != nil {
+	return err
+}
+```
+
+### Generate an OCI runtime specification
+
+```go
+// generate the spec based on the image that we pulled
+spec, err := containerd.GenerateSpec(containerd.WithImageConfig(ctx, image))
+if err != nil {
+	return err
+}
+```
+
+### Create a new container based on the image
+
+```go
+// create the container with a persistant ReadWrite layer based on the image and spec
+container, err := client.NewContainer(ctx, "redis", spec, containerd.WithNewRootFS("redis-rootfs", image))
+if err != nil {
+	return err
+}
+defer container.Delete(ctx)
+```
+
+### Run the container
+
+```go
+// use the current process's stdio
+task, err := container.NewTask(ctx, containerd.Stdio)
+if err != nil {
+	return err
+}
+defer task.Delete(ctx)
+
+pid := task.Pid()
+
+// start the redis process
+if err := task.Start(ctx); err != nil {
+	return err
+}
+
+task.Kill(ctx, syscall.SIGTERM)
+
+status, err := task.Wait(ctx)
+
+os.Exit(status)
+```
+
+This is a very simple but common usecase for containerd.  Pull an image and run a container based on that image.  The client lets you write code, use `defer` statements and any other logic you want without having to do complex filesystem operations, deal with tar files or setting up root filesystems.  The API is split up and allows you to look into the lifecycle of an executing container without the need to register hooks and provide callbacks.
+
+Over the next few weeks we will be expanding the client to support the full containerd feature set.
+
+## Testing
+
+With the initial client merged into the code base we are also starting work on integration tests.  We have full integration tests running on travis ci.  You don't have to mess with any other test frameworks to write tests for containerd, just the standard Go testing package.
+
+We also designed the client to allow mocking of calls so that our tests can be mocked out or users of our client to easily mock their own code for testing.
+
+You can view the PR [here](https://github.com/containerd/containerd/pull/910).
+
+
+## Whats Next?
+
+We still need to finish Events.  This is one of the last major features that we need before we consider containerd feature complete.  The namespace work is in progress with the initial service spec'd.
```

Comment:
There's still a few missing features on the runtime side. From the top of my head, there's no "Update" yet.